### PR TITLE
[feat] GestureDetector: add multiswipes

### PR DESCRIPTION
**Pull request by Frenzie**
This morning I finished my thought from about a year ago that I mentioned in passing last night https://github.com/koreader/koreader/pull/4601#issuecomment-464386779

I haven't quite got the GestureDetector end of it sorted out. I'll probably write it up GUI-less for now, other than an on/off switch and a first time use confirmation dialog.
<hr>
The basic idea is that you gain an infinite amount of extra gestures,
although in practice you're probably mostly limited to your 16 basic
two swipe combinations and maybe a few three swipe ones.

### CODE REVIEW

**Comment by poire-z** [comment]
Your other PR #4607 seems to include this one (?). Where do you want our comments?

File: frontend/device/gesturedetector.lua [hunk 1]
```diff
@@ -162,18 +165,24 @@ end
...
     local direction = nil
     local distance = math.sqrt(x_diff*x_diff + y_diff*y_diff)
     if x_diff ~= 0 or y_diff ~= 0 then
         local v_direction = y_diff < 0 and "north" or "south"
         local h_direction = x_diff < 0 and "west" or "east"
-        if math.abs(y_diff) > 0.577*math.abs(x_diff)
-            and math.abs(y_diff) < 1.732*math.abs(x_diff) then
+        if not simple
+           and math.abs(y_diff) > 0.577*math.abs(x_diff)
+           and math.abs(y_diff) < 1.732*math.abs(x_diff)
```
Comment (poire-z) [inline frontend/device/gesturedetector.lua]: Not your numbers I guess, but no comment says where 0.577 and 1.732 come from. Any idea?

**Comment by Frenzie** [comment]
Comments for GestureDetector here please. :-)

File: frontend/device/gesturedetector.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/device/gesturedetector.lua]: Looks like it's from the early days of the program in https://github.com/koreader/koreader/commit/15eccb4da7aa1afa44ffc0f6b8cbbb04abb521b8 but that does not, of course, answer the question.

**Comment by poire-z** [comment]
OK, but I have no more comment :) (I guess the tricky stuff is in `GestureDetector:handlePan()` , and I trust you it works well :)

File: frontend/device/gesturedetector.lua [hunk 2]
```diff
@@ -440,10 +466,11 @@ function GestureDetector:handleSwipe(tev)
     logger.dbg("swipe", swipe_direction, swipe_distance, "detected in slot", slot)
     self:clearState(slot)
     return {
-        ges = "swipe",
+        ges = ges,
         -- use first pan tev coordination as swipe start point
         pos = start_pos,
         direction = swipe_direction,
+        multiswipe_directions = multiswipe_directions,
```
Comment (Frenzie) [inline frontend/device/gesturedetector.lua]: @poire-z Perhaps this should just be `direction`. My original concept was to sneak multiswipes into regular swipes, but the kind of gestures that are broken by this are actually bugs anyway.

(E.g., for a "long diagonal swipe" you can swipe by two edges in what's now properly detected as a "multiswipe".)

Comment (Frenzie) [reply frontend/device/gesturedetector.lua]: That is, the ges = ges line was something I changed last minute before creating the PR.

**Comment by Frenzie** [comment]
Yes, that part works well with the addition of the newly added safety checks. My draft implementation from last year had some issues along these lines:



So I needed a minimum threshold, for which I needed arbitrary path detection. That all works perfectly.

File: frontend/device/gesturedetector.lua (hunk 2 above)
Comment (poire-z) [reply frontend/device/gesturedetector.lua]: Well, I don't have much of an opinion, I've never gone deep enough in the gesture code.
Question: two same-direction swipes (with a slight angle variation) still continue being a single swipe? or can you get multiswipe southeast_southeast?

Comment (Frenzie) [reply frontend/device/gesturedetector.lua]: You can give it a go with diagonal multigestures but I found it a bit error-prone. But perhaps I still had the threshold too small then.

There's the line `not pan_direction:match(msd_direction_prev)` which matches `southeast` to `south` or `east` (whichever the currently going direction is). So only something completely different will go through at all.

Then later in `if msd_direction ~= msd_direction_prev then` its' ensured that the same direction isn't added twice.

To try diagonal multigestures you'd need to get rid of the "simple" path and that :match line.

Comment (poire-z) [reply frontend/device/gesturedetector.lua]: I guess we don't really need to support diagonal multi gestures, and I somehow think it's best not to, to ease distinguish them and avoid mis-decisions.

**Comment by Frenzie** [comment]
Btw, any thoughts on this one or its follow-up @pazos ?

File: frontend/device/gesturedetector.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/device/gesturedetector.lua]: Also see [here](https://www.wolframalpha.com/input/?i=Euler%27s+constant&assumption=%7B%22C%22,+%22Euler%27s+constant%22%7D+-%3E+%7B%22NamedConstant%22,+%22dflt%22%7D&assumption=%7B%22ConstClash%22,+%22Euler%27s+constant%22%7D+-%3E+%7BE%7D) for the first. I knew it looked familiar.

The other one seems to be the result of √3.

**Comment by Frenzie** [comment]
Alright, I'll merge this tonight so the feedback can start pouring in.

**Pull request merged**